### PR TITLE
Tests: Refactoring #49 to work with latest adjacency tests/module

### LINKED ISSUE #50: ADBC driver error
## Problem

```
adbc_driver_manager.OperationalError: IO: failed to open 'Users/taddbindas/hydrofabric/v2.2/conus_nextgen.gpkg': failed to allocate memory
```

### Actual Behavior

The ADBC driver used in the `engine/` folder relies on loading the full conus gpkg into memory. On my Mac Mini (with 64GB of RAM) its failing. 

Using `sqllite3` works when using the commented version from #46 
```
(ddr) taddbindas in ~/projects/forks/ddr on refactor_tests ● λ python engine/adjacency.py ~/hydrofabric/v2.2/conus_nextgen.gpkg data/conus_adjacency.zarr
Building network graph: 100%|███████████████████████████████████████| 828289/828289 [00:00<00:00, 2073869.17it/s]
Creating sparse matrix indicies: 100%|██████████████████████████████| 828289/828289 [00:00<00:00, 2351911.75it/s]
CONUS Hydrofabric adjacency written to zarr at data/conus_adjacency.zarr
```

## Steps to Reproduce

```
(ddr) taddbindas in ~/projects/forks/ddr on refactor_tests ● λ python engine/adjacency.py -h
usage: adjacency.py [-h] pkg [path]

Create a lower triangular adjacency matrix from hydrofabric data.

positional arguments:
  pkg         Path to the hydrofabric geopackage.
  path        Path to save the zarr group. Defaults to current working directory with name appended.

options:
  -h, --help  show this help message and exit
(ddr) taddbindas in ~/projects/forks/ddr on refactor_tests ● λ python engine/adjacency.py ~/hydrofabric/v2.2/conus_nextgen.gpkg data/conus_adjacency.zarr
Traceback (most recent call last):
  File "/Users/taddbindas/projects/forks/ddr/engine/adjacency.py", line 222, in <module>
    fp = pl.read_database_uri(query=query, uri=uri, engine="adbc")
  File "/Users/taddbindas/projects/forks/ddr/.venv/lib/python3.13/site-packages/polars/io/database/functions.py", line 448, in read_database_uri
    return _read_sql_adbc(
        query,
    ...<2 lines>...
        execute_options=execute_options,
    )
  File "/Users/taddbindas/projects/forks/ddr/.venv/lib/python3.13/site-packages/polars/io/database/_utils.py", line 62, in _read_sql_adbc
    with _open_adbc_connection(connection_uri) as conn, conn.cursor() as cursor:
         ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
  File "/Users/taddbindas/projects/forks/ddr/.venv/lib/python3.13/site-packages/polars/io/database/_utils.py", line 89, in _open_adbc_connection
    return adbc_driver.connect(connection_uri)
           ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
  File "/Users/taddbindas/projects/forks/ddr/.venv/lib/python3.13/site-packages/adbc_driver_sqlite/dbapi.py", line 101, in connect
    db = adbc_driver_sqlite.connect(uri)
  File "/Users/taddbindas/projects/forks/ddr/.venv/lib/python3.13/site-packages/adbc_driver_sqlite/__init__.py", line 60, in connect
    return adbc_driver_manager.AdbcDatabase(driver=_driver_path(), uri=uri)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "adbc_driver_manager/_lib.pyx", line 501, in adbc_driver_manager._lib.AdbcDatabase.__init__
  File "adbc_driver_manager/_lib.pyx", line 260, in adbc_driver_manager._lib.check_error
adbc_driver_manager.OperationalError: IO: failed to open 'Users/taddbindas/hydrofabric/v2.2/conus_nextgen.gpkg': failed to allocate memory
```

## Proposed Solution (Optional)

Using `sqllite3` rather than adbc

### PR DESCRIPTION
**Pull request by taddyb**
## Issue Addressed

Closes #49 and makes sure code is at the HEAD
Closes #50 as conn issue is addressed

## Description

Fixes # (issue number)
- Cleans up `engine/` repo and README
- Moves `tests/` into it's own folder
- Creates a `conftest.py` file for shared fixtures

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code cleanup/refactor
- [ ] Documentation update

Other (please specify):

## Checklist

- [x] Branch is up to date with master
- [x] Updated tests or added new tests
- [ ] Tests & pre-commit hooks pass
- [ ] Updated documentation (if applicable)
- [x] Code follows established style and conventions

### CODE REVIEW

File: engine/README.md [hunk 1]
```diff
@@ -1,16 +1,15 @@
 # DDR Engine
 
 This folder contains scripts and tools meant to format versions of the hydrofabric and create objects which help our routing. Examples include:
-- Mapping MERIT streamflow predictions to the hydrofabric catchments
-- Creaing adjacency matrices for implicit muskingum cunge routing
-  - Matrices are saved in
+- Creating pyiceberg warehouses
+- Creaing adjacency matrices for implicit muskingum cunge routing. Matrices are indexed based on the CONUS geopackage and are for
```
Comment (hellkite500) [inline engine/README.md]: This ends in seeming incomplete thought...

File: engine/gages_adjacency.py [hunk 2]
```diff
@@ -305,11 +306,28 @@ def coo_to_zarr_group(
...
-    catalog = load_catalog(namespace)
-    fp = catalog.load_table("hydrofabric.flowpaths").to_polars()
-    network = catalog.load_table("hydrofabric.network").to_polars()
+    # Read hydrofabric geopackage using sqlite
+    uri = "sqlite://" + str(args.pkg)
+    query = "SELECT id,toid,tot_drainage_areasqkm FROM flowpaths"
+    # fp = pl.read_database_uri(query=query, uri=uri, engine="adbc")
+    # Using adbc is about 2 seconds faster than using the sqlite3 connection
+    conn = sqlite3.connect(args.pkg)
+    flowpaths_schema = {
```
Comment (hellkite500) [inline engine/gages_adjacency.py:317]: `flowpaths_scema` appears unused?

File: tests/test_adjacency.py [hunk 3]
```diff
@@ -8,69 +8,26 @@
...
 
 import numpy as np
 import pandas as pd
+import polars as pl
 import pytest
 import zarr
-from polars import LazyFrame, String, col
 
-from ..adjacency import coo_to_zarr, create_matrix, index_matrix  # noqa: TID252
+sys.path.insert(0, str(Path(__file__).parents[1] / "engine"))
```
Comment (hellkite500) [inline tests/test_adjacency.py:21]: This is a bit ugly...

**Review by hellkite500 — COMMENTED** [review]
I'm not sure that the refactored test fixtures capture the same kinds of edge cases with missing/terminal components (at least not explicitly) that were being tested previously for the adjacency functionality.  I need to think about this a little more, but I'm leaning towards ensuring a couple simple fixtures exist with intentionally designed missing/terminal points to ensure those cases are properly tested.

File: engine/gages_adjacency.py (hunk 2 above)
Comment (taddyb) [reply engine/gages_adjacency.py:317]: Thanks. This should be used below with `schema_overrides`

File: tests/test_adjacency.py (hunk 3 above)
Comment (taddyb) [reply tests/test_adjacency.py:21]: Not sure how else to import the modules from `adjacency` and `gages_adjacency` given tests/ is at the project root. 

another option would be to make `engine/` it's own installable package, however since the code within are scripts with argparse endpoints that doesn't make much sense to me

We could make the `engine/` repo more complex, but this single line of inserting the `engine/` folder into the system path seems to work well for right now

File: engine/README.md (hunk 1 above)
Comment (taddyb) [reply engine/README.md]: Addressed

**Comment by taddyb** [comment]
@hellkite500 Thanks for the review

> I'm not sure that the refactored test fixtures capture the same kinds of edge cases with missing/terminal components (at least not explicitly) that were being tested previously for the adjacency functionality. 

I ended up scrapping any of the refactored test fixtures that I changed in #49 and am using the ones you added in #46. The only modifications were adding a new row for total drainage areas in fp and hl_uri in network. See below for the example

```python
_flowpath_table_schema = {
    "id": pl.String,
    "toid": pl.String,
    "tot_drainage_areasqkm": pl.Float64,
}

@pytest.fixture
def simple_flowpaths() -> pl.LazyFrame:
    """Create a simple flowpaths LazyFrame for testing."""
    data = {
        "id": ["wb-1", "wb-2"],
        "toid": ["nex-1", "nex-1"],
        "tot_drainage_areasqkm": [60, 120],
    }
    fp = pl.LazyFrame(data, schema=_flowpath_table_schema)
    return fp

_network_table_schema = {
    "id": pl.String,
    "toid": pl.String,
    "hl_uri": pl.String,
}


@pytest.fixture
def simple_network() -> pl.LazyFrame:
    """Create a simple network LazyFrame for testing."""
    data = {
        "id": ["nex-1", "wb-1", "wb-2"],
        "toid": [None, "nex-1", "nex-1"],  # Use None for null values
        "hl_uri": [None, "gages-01234567", "gages-01234567"],
    }
    network = pl.LazyFrame(data, schema=_network_table_schema)
    return network
```

This should address your concern.


> I need to think about this a little more, but I'm leaning towards ensuring a couple simple fixtures exist with intentionally designed missing/terminal points to ensure those cases are properly tested.

The only way I'm doing this in `test_gages_adjacency.py` right now is through this fixture below where I'm looking up a gage that doesn't exist. The way I designed the gage subsetting code it to create subsets after a CONUS adjacency is built. Thus, I think the intentionally missing terminal points should be tested against the CONUS scripts and are out of the scope of this PR.
```
@pytest.fixture
def non_existing_gage():
    """Creates a gauge not in the testing fixtures"""
    return Gauge(STAID="0000", DRAIN_SQKM=123.4)
```

File: engine/README.md [hunk 4]
```diff
@@ -1,16 +1,23 @@
...
 
 This folder contains scripts and tools meant to format versions of the hydrofabric and create objects which help our routing. Examples include:
-- Mapping MERIT streamflow predictions to the hydrofabric catchments
-- Creaing adjacency matrices for implicit muskingum cunge routing
-  - Matrices are saved in
+- Creaing adjacency matrices for implicit muskingum cunge routing. Matrices are indexed based on the CONUS geopackage.
+- Creating pyiceberg warehouses for data structure storage
 
-How to run adjacency matrix:
+To install these dependencies, please run the following command from the project root
```
Comment (hellkite500) [inline engine/README.md:7]: So polite =)

File: engine/gages_adjacency.py [hunk 5]
```diff
@@ -70,7 +73,7 @@ def find_origin(gauge: Gauge, fp: pl.LazyFrame, network: pl.LazyFrame) -> np.nda
         raise ValueError from e
 
 
-def subset(origin: str, wb_network_dict: dict[str, list[str]]) -> list[str, str]:
+def subset(origin: str, wb_network_dict: dict[str, list[str]]) -> list[tuple[str]]:
```
Comment (hellkite500) [inline engine/gages_adjacency.py]:
```suggestion
def subset(origin: str, wb_network_dict: dict[str, list[str]]) -> list[tuple[str,str]]:
```

File: engine/gages_adjacency.py [hunk 6]
```diff
@@ -82,7 +85,7 @@ def subset(origin: str, wb_network_dict: dict[str, list[str]]) -> list[str, str]
 
     Returns
     -------
-    list[str, str]
+    list[tuple[str]]
```
Comment (hellkite500) [inline engine/gages_adjacency.py]:
```suggestion
    list[tuple[str, str]]
```

File: engine/gages_adjacency.py [hunk 7]
```diff
@@ -108,13 +111,13 @@ def trace_upstream_recursive(current_id: str) -> None:
 
 
 def create_coo(
-    connections: list[str, str], conus_mapping: dict[str, int]
+    connections: list[tuple[str]], conus_mapping: dict[str, int]
```
Comment (hellkite500) [inline engine/gages_adjacency.py]:
```suggestion
    connections: list[tuple[str, str]], conus_mapping: dict[str, int]
```

File: engine/gages_adjacency.py [hunk 8]
```diff
@@ -108,13 +111,13 @@ def trace_upstream_recursive(current_id: str) -> None:
...
 ) -> tuple[sparse.coo_matrix, list[str]]:
     """A function to create a coo matrix out of the ts_ordering from the conus_adjacency matrix indices
 
     Parameters
     ----------
-    connections: list[str, str]
+    connections: list[tuple[str]]
```
Comment (hellkite500) [inline engine/gages_adjacency.py]:
```suggestion
    connections: list[tuple[str, str]]
```

File: engine/gages_adjacency.py [hunk 9]
```diff
@@ -305,11 +308,28 @@ def coo_to_zarr_group(
     else:
         raise FileNotFoundError("Can't find the Gauge Information file")
 
-    # Read in hydrofabric
-    namespace = "hydrofabric"
-    catalog = load_catalog(namespace)
-    fp = catalog.load_table("hydrofabric.flowpaths").to_polars()
-    network = catalog.load_table("hydrofabric.network").to_polars()
+    # Read hydrofabric geopackage using sqlite
+    uri = "sqlite://" + str(args.pkg)
```
Comment (hellkite500) [inline engine/gages_adjacency.py]:
```suggestion
    # uri = "sqlite://" + str(args.pkg)
```

File: tests/test_adjacency.py (hunk 3 above)
Comment (hellkite500) [reply tests/test_adjacency.py:21]: > another option would be to make engine/ it's own installable package, however since the code within are scripts with argparse endpoints that doesn't make much sense to me

I would propose this long term.  Perhaps a namesapce package would be worthwhile in this project so you could have things like

```
ddr.engine
```
and
```
ddr.routing
```
or some other organization.

But ultimately, the refactor in `adjancency.py` was intended to make the functionality within more accessible as a module so other scripts could be developed which may want to create adjacency matricies inline instead of having to read/write them from disk in a higher order pipeline.

Also, having the callable modules in a package is helpful for distribution, they can be installed via python package managers and executed from an environment conveniently, think about something like `pytest` which does just that -- it provides code to build upon, but also an executable entrypoint for running a standard entrypoint.

**Review by hellkite500 — COMMENTED** [review]
I would suggest looking at the typing suggestions before merging, but otherwise I would say this is pretty close to workable

File: tests/test_adjacency.py (hunk 3 above)
Comment (taddyb) [reply tests/test_adjacency.py:21]: If you're fine with breaking up the work then that's fine with me. It's outside the scope of this PR so I opened #58 to track the issue.
